src/Akka.Streams.Kafka/Stages/Consumers/Abstract/ExternalSingleSourceLogic.cs

```diff
@@ -1,7 +1,38 @@
+using System;
+using System.Threading.Tasks;
+using Akka.Actor;
+using Akka.Streams.Kafka.Settings;
+using Akka.Streams.Kafka.Stages.Consumers.Actors;
+
 namespace Akka.Streams.Kafka.Stages.Consumers.Abstract
 {
-    public class ExternalSingleSourceLogic
+    /// <summary>
+    /// Single source logic for externally provided <see cref="KafkaConsumerActor{K,V}"/>
+    /// </summary>
+    /// <typeparam name="K">Key type</typeparam>
+    /// <typeparam name="V">Value type</typeparam>
+    /// <typeparam name="TMessage">Message type</typeparam>
+    internal class ExternalSingleSourceLogic<K, V, TMessage> : BaseSingleSourceLogic<K, V, TMessage>
     {
-        // TODO
+        private readonly IActorRef _consumerActor;
+        private readonly IManualSubscription _subscription;
+
+        public ExternalSingleSourceLogic(SourceShape<TMessage> shape, IActorRef consumerActor, IManualSubscription subscription,
+                                         TaskCompletionSource<NotUsed> completion, Attributes attributes, 
+                                         Func<BaseSingleSourceLogic<K, V, TMessage>, IMessageBuilder<K, V, TMessage>> messageBuilderFactory) 
+            : base(shape, completion, attributes, messageBuilderFactory)
+        {
+            _consumerActor = consumerActor;
+            _subscription = subscription;
+        }
+
+        /// <inheritdoc />
+        protected override IActorRef CreateConsumerActor() => _consumerActor;
```

Comment:
I'm sure I'm missing something, but where does the output for this get consumed?

Reply:
No problem.
This `ExternalSingleSourceLogic` class is derived from `BaseSingleSourceLogic`, which contains all message consuming behavior. Child classes just specify, how to get consumer actor reference (usually create own actor or reuse existing, like here), and then base class requests messages from this actor and handles his responses.
Is that what you were asking about?

Reply:
Ah yes, thanks - just wanted to have the big picture for when I review this later. Seems simple enough.